packages/sdk/src/protocols/filter/subscription_manager.ts

```diff
@@ -0,0 +1,344 @@
+import type { Peer, PeerId } from "@libp2p/interface";
+import { FilterCore } from "@waku/core";
+import {
+  type Callback,
+  type ContentTopic,
+  type CoreProtocolResult,
+  type IDecodedMessage,
+  type IDecoder,
+  IProtoMessage,
+  type ISubscriptionSDK,
+  type PeerIdStr,
+  ProtocolError,
+  type PubsubTopic,
+  type SDKProtocolResult,
+  type SubscribeOptions,
+  SubscriptionCallback
+} from "@waku/interfaces";
+import { WakuMessage } from "@waku/proto";
+import { groupByContentTopic, Logger } from "@waku/utils";
+
+import { DEFAULT_KEEP_ALIVE, DEFAULT_SUBSCRIBE_OPTIONS } from "./constants";
+import { ReliabilityMonitor } from "./reliability_monitor";
+
+const DEFAULT_MAX_PINGS = 3;
+
+const log = new Logger("subscription-manager");
+
+export class SubscriptionManager implements ISubscriptionSDK {
+  private keepAliveTimer: number | null = null;
+  private peerFailures: Map<string, number> = new Map();
+  private maxPingFailures: number = DEFAULT_MAX_PINGS;
+
+  private reliabilityMonitor: ReliabilityMonitor;
+
+  private subscriptionCallbacks: Map<
+    ContentTopic,
+    SubscriptionCallback<IDecodedMessage>
+  >;
+
+  public constructor(
+    private readonly pubsubTopic: PubsubTopic,
+    private protocol: FilterCore,
+    private getPeers: () => Peer[],
+    private readonly renewPeer: (peerToDisconnect: PeerId) => Promise<Peer>
+  ) {
+    this.pubsubTopic = pubsubTopic;
+    this.subscriptionCallbacks = new Map();
+
+    this.reliabilityMonitor = new ReliabilityMonitor(
```

Comment:
based on the original design - we'd need to use it in `lightPush` too
or do you have another idea how to make it happen?